cibyl/utils/files.py

```diff
@@ -16,10 +16,95 @@
 import logging
 import os
 from os import PathLike
+from pathlib import Path
 
 LOG = logging.getLogger(__name__)
 
 
+class FileSearch:
+    """Allows for complex search queries targeting files on the filesystem.
+    """
+
+    def __init__(self, directory):
+        """Constructor.
+
+        :param directory: Path to directory to look for files in.
+        :type directory: str
+        """
+        self._directory = directory
+        self._recursive = False
+        self._extensions = []
+
+    def with_recursion(self):
```

Comment:
Out of curiosity, why not have the _recursive and _extension attributes in the constructor? I've seen this pattern before and I don't quite see the advantages

Reply:
Not exactly, but this is similar to the Builder pattern. The idea behind it is that it is easier to compose and extend the file search. For example, if tomorrow I want another filter for a file name, I will have to add a new 'with_name' function and update 'get' with the condition. I have extended the interface instead of modifying it, so anyone who is using the class already will not notice the change.